<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<hr style="height:.9px;border:none;color:#333;background-color:#333;" />

<h2>Homework - Hyperparameter Tuning</h2>
<h4>DAT-5303 | Machine Learning</h4>

<br>
<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<hr style="height:.9px;border:none;color:#333;background-color:#333;" />

**Team 14 :** 
- Zaynab Zennour
- Sreekar Bathula
- Olubiyi George
- Deena Jivan
- Akoh Jackson Udeng


<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<br>
The models we will use : 

* Random Forest
* Gradient Boosted Machine (GBM)

<h3>Step 1: Imports</h3>


In [14]:
# importing critical libraries
import pandas            as pd
import matplotlib.pyplot as plt
import seaborn           as sns
import itertools
import numpy as np 


# importing model types
import sklearn.linear_model                            # linear models 
from sklearn.tree     import DecisionTreeRegressor     # regression trees
from sklearn.ensemble import RandomForestRegressor     # random forest 
from sklearn.ensemble import GradientBoostingRegressor # GBM
from sklearn.metrics import r2_score



# importing ML tools
from sklearn.model_selection import train_test_split   
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV

# loading data and setting display options
housing = pd.read_excel('./__datasets/housing_feature_rich.xlsx')


# setting pandas print options
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 100)

In [3]:
x_variables = ['Garage_Cars', 'Overall_Qual', 'Total_Bsmt_SF',
               'NridgHt', 'Kitchen_AbvGr', 'has_Second_Flr',
               'Mas_Vnr_Area', 'has_Garage', 'Porch_Area',
               'NWAmes', 'OldTown', 'Overall_Cond',
               'Edwards', 'Somerst', 'Fireplaces',
               'Second_Flr_SF', 'First_Flr_SF', 'has_Mas_Vnr',
               'CulDSac', 'Total_Bath', 'Crawfor', 'Garage_Area',
               'has_Porch']


full_x = ['Overall_Qual', 'Overall_Cond', 'Mas_Vnr_Area', 'Total_Bsmt_SF',
          'First_Flr_SF', 'Second_Flr_SF', 'Gr_Liv_Area', 'Full_Bath',
          'Half_Bath', 'Kitchen_AbvGr', 'TotRms_AbvGr', 'Fireplaces',
          'Garage_Cars', 'Garage_Area', 'Porch_Area', 'log_Lot_Area',
          'has_Second_Flr', 'has_Garage', 'has_Mas_Vnr', 'has_Porch',
          'Total_Bath', 'CulDSac', 'BrkSide', 'CollgCr', 'Crawfor',
          'Edwards', 'Gilbert', 'Mitchel', 'NWAmes', 'NridgHt', 'OldTown',
          'Sawyer', 'SawyerW', 'Somerst', 'Other_NH']


reduced_x = ['Overall_Qual', 'Gr_Liv_Area', 'Full_Bath',
             'Kitchen_AbvGr', 'TotRms_AbvGr', 'Fireplaces',
             'Garage_Cars', 'Garage_Area', 'Porch_Area', 
             'log_Lot_Area', 'has_Second_Flr', 'has_Garage',
             'has_Mas_Vnr', 'has_Porch', 'Total_Bath', 'CulDSac']

housing['log_Sale_Price'] = np.log(housing['Sale_Price'])

########################################
# checking results
########################################
housing.head(n = 5)
# preparing x-variables
# x_data = housing.loc[ : , reduced_x ] # x-variables (can change this)
x_data = housing.drop(['Sale_Price', 'log_Sale_Price'], axis = 1)


# preparing y-variable
#y_data = housing.loc[ : , 'Sale_Price']    # y-variable
y_data = housing.loc[ : , 'log_Sale_Price'] # y-variable




<br>

The models we will use for tuning are :
- Random Forest
- Gradient Boosted Machine (GBM)




<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 2: Train-Test Split</h3>

In [4]:
# train-test split with stratification
X_train, X_test, y_train, y_test = train_test_split(
            x_data,
            y_data,
            test_size    = 0.25,
            random_state = 219 )




<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 3: Checking Available Hyperparameters</h3>


In [ ]:
# Calling help() on each of our models : GradientBoostingRegressor
help(GradientBoostingRegressor)

In [ ]:
# Calling help() on each of our models : RandomForestRegressor
help(RandomForestRegressor)

<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 4: Model Development - Default Hyperparameters</h3>

- **Random Forest Regressor** : Using default Hyperparameters and results 

In [5]:
# Define the default hyperparameters for Random Forest (1)
rf_default_params_1 = {'n_estimators': 100, 'max_depth': 2, 'min_samples_leaf': 5}
rf_default_1 = RandomForestRegressor(**rf_default_params_1)
rf_default_1.fit(X_train, y_train)
rf_default_train_score_1 = r2_score(y_train, rf_default_1.predict(X_train)).round(3)
rf_default_test_score_1 = r2_score(y_test, rf_default_1.predict(X_test)).round(3)

# Define the default hyperparameters for Random Forest (2)
rf_default_params_2 = {'n_estimators': 200, 'max_depth': 3,  'min_samples_leaf': 4}
rf_default_2 = RandomForestRegressor(**rf_default_params_2)
rf_default_2.fit(X_train, y_train)
rf_default_train_score_2 = r2_score(y_train, rf_default_2.predict(X_train)).round(3)
rf_default_test_score_2 = r2_score(y_test, rf_default_2.predict(X_test)).round(3)


# Define the default hyperparameters for Random Forest (3)
rf_default_params_3 = {'n_estimators': 300, 'max_depth': 5, 'min_samples_leaf': 3}
rf_default_3 = RandomForestRegressor(**rf_default_params_3)
rf_default_3.fit(X_train, y_train)
rf_default_train_score_3 = r2_score(y_train, rf_default_3.predict(X_train)).round(3)
rf_default_test_score_3 = r2_score(y_test, rf_default_3.predict(X_test)).round(3)

# Define a dictionary to store the results
results_dict = {
    'Model': ['Random Forest (default 1)', 'Random Forest (default 2)', 'Random Forest (default 3)'],
    'n_estimators': [rf_default_params_1['n_estimators'], rf_default_params_2['n_estimators'], rf_default_params_3['n_estimators']],
    'max_depth': [rf_default_params_1['max_depth'], rf_default_params_2['max_depth'], rf_default_params_3['max_depth']],
    'min_samples_leaf': [rf_default_params_1['min_samples_leaf'], rf_default_params_2['min_samples_leaf'], rf_default_params_3['min_samples_leaf']],
    'Train R-Square': [rf_default_train_score_1, rf_default_train_score_2, rf_default_train_score_3],
    'Test R-Square': [rf_default_test_score_1, rf_default_test_score_2, rf_default_test_score_3]
}

# Create a Pandas DataFrame from the results dictionary
results_df = pd.DataFrame(results_dict)

# Print the results in tabular format
print(results_df.to_string(index=False))

                    Model  n_estimators  max_depth  min_samples_leaf  Train R-Square  Test R-Square
Random Forest (default 1)           100          2                 5           0.641          0.640
Random Forest (default 2)           200          3                 4           0.765          0.753
Random Forest (default 3)           300          5                 3           0.867          0.831


- **Gradient Boosting Regressor** : Using default Hyperparameters and results 

In [6]:

# Default hyperparameters for Gradient Boosted Machine (GBM) (1)
gbm_default_params_1 = {'n_estimators': 100, 'max_depth': 3, 'min_samples_split': 4}
gbm_default_1 = GradientBoostingRegressor(**gbm_default_params_1)
gbm_default_1.fit(X_train, y_train)
gbm_default_train_score_1 = r2_score(y_train, gbm_default_1.predict(X_train)).round(3)
gbm_default_test_score_1 = r2_score(y_test, gbm_default_1.predict(X_test)).round(3)

# Default hyperparameters for Gradient Boosted Machine (GBM) (2)
gbm_default_params_2 = {'n_estimators': 200, 'max_depth': 4, 'min_samples_split': 2}
gbm_default_2 = GradientBoostingRegressor(**gbm_default_params_2)
gbm_default_2.fit(X_train, y_train)
gbm_default_train_score_2 = r2_score(y_train, gbm_default_2.predict(X_train)).round(3)
gbm_default_test_score_2 = r2_score(y_test, gbm_default_2.predict(X_test)).round(3)

# Default hyperparameters for Gradient Boosted Machine (GBM) (3)
gbm_default_params_3 = {'n_estimators': 300, 'max_depth': 5, 'min_samples_split': 6}
gbm_default_3 = GradientBoostingRegressor(**gbm_default_params_3)
gbm_default_3.fit(X_train, y_train)
gbm_default_train_score_3 = r2_score(y_train, gbm_default_3.predict(X_train)).round(3)
gbm_default_test_score_3 = r2_score(y_test, gbm_default_3.predict(X_test)).round(3)


# Define a dictionary to store the results
results_dict = {
    'Model': ['Gradient Boosted Machine (default 1)', 'Gradient Boosted Machine (default 2)', 'Gradient Boosted Machine (default 3)'],
    'n_estimators': [gbm_default_params_1['n_estimators'], gbm_default_params_2['n_estimators'], gbm_default_params_3['n_estimators']],
    'max_depth': [gbm_default_params_1['max_depth'], gbm_default_params_2['max_depth'], gbm_default_params_3['max_depth']],
    'min_samples_split': [gbm_default_params_1['min_samples_split'], gbm_default_params_2['min_samples_split'], gbm_default_params_3['min_samples_split']],
    'Train R-Square': [gbm_default_train_score_1, gbm_default_train_score_2, gbm_default_train_score_3],
    'Test R-Square': [gbm_default_test_score_1, gbm_default_test_score_2, gbm_default_test_score_3]
}

# Create a Pandas DataFrame from the results dictionary
results_df = pd.DataFrame(results_dict)

# Print the results in tabular format
print(results_df.to_string(index=False))

                               Model  n_estimators  max_depth  min_samples_split  Train R-Square  Test R-Square
Gradient Boosted Machine (default 1)           100          3                  4           0.931          0.883
Gradient Boosted Machine (default 2)           200          4                  2           0.968          0.886
Gradient Boosted Machine (default 3)           300          5                  6           0.990          0.882


<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 5: Hyperparameter Tuning Part I</h3><br>
<br><br>
<strong>Requirement:</strong>  at least three hyperparameters per model.<br>
<strong>Recommendation:</strong> Keep the number of models (i.e., iterations) you are building to:

* Less than 2,000 for Random Forest and GBM  (not including cross-validation)
* Less than 10,000 for all other model types (not including cross-validation)


In [19]:
# Define ranges for each of the hyperparameters we would like tune. 
# this code calculates how many models we are building (based on the combinations of hyperparameters we are tuning).
# parameters grid for RF (option 1) 
criterion_range = ["squared_error", "absolute_error", "poisson" ]
split_range  = [ 1, 5, 10, 15, 20]
depth_range     = range(5, 30, 10)
leaf_range      = range(1, 50 ,2)   

#define the product of possible combinations for each parameter 
params_grid = itertools.product( depth_range, split_range, leaf_range, criterion_range)

num_models = len(list(params_grid))
print(f"Number of Random Forest models to be built: {num_models}")


Number of Random Forest models to be built: 1125


In [39]:
# parameters grid for RF (option 2) 
criterion_range = ["squared_error" ]
split_range  = [ 1, 5, 10, 15, 20]
depth_range     = range(5, 100, 5)
leaf_range      = range(1, 30 ,2)

params_grid = itertools.product( depth_range, split_range, leaf_range, criterion_range)

num_models = len(list(params_grid))
print(f"Number of Random Forest models to be built: {num_models}")


Number of Random Forest models to be built: 1425


In [20]:
# Define ranges for each of the hyperparameters we would like tune. 
# this code calculates how many models we are building (based on the combinations of hyperparameters we are tuning).
criterion_range = ['friedman_mse', 'squared_error', 'mse' ]
split_range  = [2, 5, 10, 15]
depth_range     = range(2, 6 ,1)  
leaf_range      = range(2, 60 ,2)   

#define the product of possible combinations for each parameter 
params_grid = itertools.product( depth_range, split_range, leaf_range, criterion_range)

num_models = len(list(params_grid))
print(f"Number of Gradient Boosting Regressor models to be built: {num_models}")


Number of Gradient Boosting Regressor models to be built: 1392


<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 6: Hyperparameter Tuning Part II</h3><br>


In [40]:
# declaring a hyperparameter space
criterion_range = ["squared_error", "absolute_error", "poisson" ]
split_range  = [ 1, 5, 10, 15, 20]
depth_range     = range(5, 30, 10)
leaf_range      = range(1, 50 ,2)   



# creating a hyperparameter grid for RF

param_grid_rf = {'criterion' : criterion_range,
                 'min_samples_split' : split_range,
                 'max_depth' : depth_range,
                 'min_samples_leaf' : leaf_range
                 }

# declaring a hyperparameter space
# hyperparameter grid for GBM 

criterion = ['friedman_mse', 'squared_error', 'mse' ]
split  = [2, 5, 10, 15]
depth     = range(2, 6 ,1)  
leaf      = range(2, 60 ,2)   



param_grid_gbm   = {'criterion' : criterion,
                    'min_samples_split' : split,
                    'max_depth' : depth,
                    'min_samples_leaf' : leaf }



In [ ]:
# declaring a hyperparameter space
criterion_range = ["squared_error", "absolute_error", "poisson" ]
split_range  = [10, 15]
depth_range     = range(5, 30, 10)
leaf_range      = range(1, 60 ,2) 

# creating a hyperparameter grid for RF

param_grid_rf = {'criterion' : criterion_range,
                 'min_samples_split' : split_range,
                 'max_depth' : depth_range,
                 'min_samples_leaf' : leaf_range
                 }

# declaring a hyperparameter space
# hyperparameter grid for GBM 

criterion = ['friedman_mse', 'squared_error', 'mse' ]
split  = [2, 5, 10, 15]
depth     = range(2, 6 ,1)  
leaf      = range(2, 60 ,2)   



param_grid_gbm   = {'criterion' : criterion,
                    'min_samples_split' : split,
                    'max_depth' : depth,
                    'min_samples_leaf' : leaf }



<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 7: Hyperparameter Tuning Part III</h3><br>

Approximately, the processing time varies from 5 min to 30 min when the number of combinations gets higher (1000 and above).


**GradientBoostingRegressor Fine Tunning Process** 

In [ ]:

# INSTANTIATING the model object without hyperparameters
model_gbm = GradientBoostingRegressor(random_state = 219)



# RandomizedSearchCV object
model_cv_gbm = RandomizedSearchCV(estimator           = model_gbm,
                              param_distributions = param_grid_gbm,
                              n_iter              = 1000,
                              cv                  = 3,
                              random_state        = 219)



# FITTING to the FULL DATASET (due to cross-validation)
model_cv.fit(x_data, y_data)



# printing the optimal parameters and best score
print(f"Tuned Parameters: {model_cv_gbm.best_params_}")
print(f"Tuned R-Square:   {model_cv_gbm.best_score_}")

**RandomForestRegressor Fine Tunning Process**

In [41]:

# INSTANTIATING the model object without hyperparameters
model_rf = RandomForestRegressor(random_state = 219)



# RandomizedSearchCV object
model_cv_rf = RandomizedSearchCV(estimator           = model_rf,
                              param_distributions = param_grid_rf,
                              n_iter              = 20,
                              cv                  = 5,
                              random_state        = 219)



# FITTING to the FULL DATASET (due to cross-validation)
model_cv_rf.fit(x_data, y_data)



# printing the optimal parameters and best score
print(f"Tuned Parameters: {model_cv_rf.best_params_}")
print(f"Tuned R-Square:   {model_cv_rf.best_score_}")



C:\Users\LENOVO\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
25 fits failed out of a total of 100.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\LENOVO\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\LENOVO\anaconda3\lib\site-packages\sklearn\ensemble\_forest.py", line 450, in fit
    trees = Parallel(
  File "C:\Users\LENOVO\anaconda3\lib\site-packages\joblib\parallel.py", line 1043, in __call__
    if self.dispatch_one_batch(iterator):
  File "C:\Users\LENOVO\anaconda3\lib\site-packa

Tuned Parameters: {'min_samples_split': 10, 'min_samples_leaf': 1, 'max_depth': 40, 'criterion': 'squared_error'}
Tuned R-Square:   0.8631417429230112


- Exploring the hyperparameter tuning results for one of our best tuned models (for RF and GBM). 

- Looking for hyperparameter combinations that tied for first place. 

In [13]:
# building a model based on hyperparameter tuning results

# INSTANTIATING a model with tuned values
model = RandomForestRegressor(min_samples_split = 5, 
                              min_samples_leaf = 3, 
                              max_depth = 30, 
                              criterion = "squared_error", 
                              random_state     = 219)


# FITTING to the TRAINING data
model_fit = model.fit(X_train, y_train)


# PREDICTING based on the testing set
model_pred = model.predict(X_test)


# SCORING the results
model_train_score = model.score(X_train, y_train).round(4) # using R-square
model_test_score  = model.score(X_test, y_test).round(4)   # using R-square
model_gap         = abs(model_train_score - model_test_score).round(4)


# displaying results
print('Training Score :', model_train_score)
print('Testing Score  :', model_test_score)
print('Train-Test Gap :', model_gap)

Training Score : 0.9567
Testing Score  : 0.8676
Train-Test Gap : 0.0891


- Test the best 3 tuned RF models 

In [44]:
# Define the BEST parameters for Random Forest Top1
rf_default_params_1 = {'criterion': 'squared_error', 'max_depth': 60, 'min_samples_leaf': 2, 'random_state' :219, 'min_samples_split' : 3}

                            
rf_default_1 = RandomForestRegressor(**rf_default_params_1)
rf_default_1.fit(X_train, y_train)
rf_default_train_score_1 = r2_score(y_train, rf_default_1.predict(X_train)).round(3)
rf_default_test_score_1 = r2_score(y_test, rf_default_1.predict(X_test)).round(3)

# Define the BEST parameters for Random Forest Top2
rf_default_params_2 = {'criterion': 'squared_error', 'max_depth': 30, 
                       'min_samples_leaf': 3, 'random_state' :219, 'min_samples_split' : 5}

rf_default_2 = RandomForestRegressor(**rf_default_params_2)
rf_default_2.fit(X_train, y_train)
rf_default_train_score_2 = r2_score(y_train, rf_default_2.predict(X_train)).round(3)
rf_default_test_score_2 = r2_score(y_test, rf_default_2.predict(X_test)).round(3)


# Define the BEST parameters for Random Forest Top3
rf_default_params_3 = {'random_state' :219,
                      'min_samples_split': 10, 'min_samples_leaf': 1, 'max_depth': 40, 'criterion': 'squared_error'}

rf_default_3 = RandomForestRegressor(**rf_default_params_3)
rf_default_3.fit(X_train, y_train)
rf_default_train_score_3 = r2_score(y_train, rf_default_3.predict(X_train)).round(3)
rf_default_test_score_3 = r2_score(y_test, rf_default_3.predict(X_test)).round(3)

# Define a dictionary to store the results
results_dict = {
    'Model': ['Random Forest (Top 1)', 'Random Forest (Top 2)', 'Random Forest (Top 3)'],
    'criterion': [rf_default_params_1['criterion'], rf_default_params_2['criterion'], rf_default_params_3['criterion']],
    'max_depth': [rf_default_params_1['max_depth'], rf_default_params_2['max_depth'], rf_default_params_3['max_depth']],
    'min_samples_split': [rf_default_params_1['min_samples_split'], rf_default_params_2['min_samples_split'], rf_default_params_3['min_samples_split']],
    'min_samples_leaf': [rf_default_params_1['min_samples_leaf'], rf_default_params_2['min_samples_leaf'], rf_default_params_3['min_samples_leaf']],
    'Train R-Square': [rf_default_train_score_1, rf_default_train_score_2, rf_default_train_score_3],
    'Test R-Square': [rf_default_test_score_1, rf_default_test_score_2, rf_default_test_score_3]
}

# Create a Pandas DataFrame from the results dictionary
results_df = pd.DataFrame(results_dict)

# Print the results in tabular format
print(results_df.to_string(index=False))

                Model     criterion  max_depth  min_samples_split  min_samples_leaf  Train R-Square  Test R-Square
Random Forest (Top 1) squared_error         60                  3                 2           0.970          0.869
Random Forest (Top 2) squared_error         30                  5                 3           0.957          0.868
Random Forest (Top 3) squared_error         40                 10                 1           0.961          0.869


In [22]:
# building a model based on hyperparameter tuning results

# INSTANTIATING a model with tuned values
model = GradientBoostingRegressor(min_samples_split = 2, 
                              min_samples_leaf = 2, 
                              max_depth = 4, 
                              criterion = "friedman_mse", 
                              random_state     = 219)


# FITTING to the TRAINING data
model_fit = model.fit(X_train, y_train)


# PREDICTING based on the testing set
model_pred = model.predict(X_test)


# SCORING the results
model_train_score = model.score(X_train, y_train).round(4) # using R-square
model_test_score  = model.score(X_test, y_test).round(4)   # using R-square
model_gap         = abs(model_train_score - model_test_score).round(4)


# displaying results
print('Training Score :', model_train_score)
print('Testing Score  :', model_test_score)
print('Train-Test Gap :', model_gap)

Training Score : 0.9511
Testing Score  : 0.8789
Train-Test Gap : 0.0722


- Display the top three models for RF

In [11]:
# A function that displays the best hypertuning results and their models' test Rsquare
# Source : Pr. Chase Kusterer, Script 5 - Hyperparameter Tuning - Solution, DAT-5390 | Computational Data Analytics with Python
def tuning_results(cv_results, n=5):
    """
This function will display the top "n" models from hyperparameter tuning,
based on "rank_test_score".

PARAMETERS
----------
cv_results = results dictionary from the attribute ".cv_results_"
n          = number of models to display
    """
    param_lst = []

    for result in cv_results["params"]:
        result = str(result).replace(":", "=")
        param_lst.append(result[1:-1])


    results_df = pd.DataFrame(data = {
        "Model_Rank" : cv_results["rank_test_score"],
        "Mean_Test_Score" : cv_results["mean_test_score"],
        "SD_Test_Score" : cv_results["std_test_score"],
        "Parameters" : param_lst
    })


    results_df = results_df.sort_values(by = "Model_Rank", axis = 0)
    return results_df.head(n = n)



In [42]:

# run tuning_results() on the hyperparameter tuning results, displaying the top 3 best models for RF 
tuning_results (n=3, cv_results = model_cv_rf.cv_results_ )

Model_Rank  Mean_Test_Score  SD_Test_Score                                                                                     Parameters
15           1         0.863142       0.015025  'min_samples_split'= 10, 'min_samples_leaf'= 1, 'max_depth'= 40, 'criterion'= 'squared_error'
6            2         0.860115       0.017050   'min_samples_split'= 5, 'min_samples_leaf'= 3, 'max_depth'= 20, 'criterion'= 'squared_error'
2            3         0.858830       0.017018  'min_samples_split'= 10, 'min_samples_leaf'= 3, 'max_depth'= 85, 'criterion'= 'squared_error'

- The best parameters for **Random Forest Regressor** (Train score is 0.96 and Test score is 0.869) are : 

      {'criterion' : 'squared_error', 'max_depth' : 40, 'min_samples_leaf' : 1, 'min_samples_split' : 10}



<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3>Step 8: (optional) Explore Tuned Hyperparameters</h3><br>
After tuning, it is a great idea to explore any hyperparameters that were markedly different from their default values once tuned. Investigate the purpose of these hyperparameters in <a href="https://scikit-learn.org/stable/index.html">the scikit-learn documentation</a>.

<hr style="height:.9px;border:none;color:#333;background-color:#333;" />
<h3> References:</h3><br>

- Pr. Chase Kusterer, Script 5 - Hyperparameter Tuning - Solution, DAT-5390 | Computational Data Analytics with Python